### PR TITLE
adding cov report to lint command

### PR DESCRIPTION
**Pull request by esharf**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
related to: https://github.com/demisto/etc/issues/22899

## Description
Adding coverage report to demisto-sdk lint command.
[content coverage](https://375905-60525392-gh.circle-artifacts.com/0/artifacts/coverage_report/html/index.html)
FYI @glicht @anara123

## Must have
- [x] Tests
- [ ] Documentation

### CODE REVIEW

File: demisto_sdk/commands/lint/helpers.py [hunk 1]
```diff
@@ -457,3 +459,48 @@ def split_warnings_errors(output: str):
             other_msg_list.append(msg)
 
     return error_list, warnings_list, other_msg_list
+
+
+def coverage_report_editor(coverage_file, code_file_absolute_path):
```
Comment (yaakovi) [inline demisto_sdk/commands/lint/helpers.py:464]: add docstring here

File: demisto_sdk/commands/lint/helpers.py [hunk 2]
```diff
@@ -457,3 +459,48 @@ def split_warnings_errors(output: str):
...
+
+def coverage_files():
+    for pack in Path('Packs').rglob('*'):
+        for code_type in ['Integrations', 'Scripts']:
+            code_dirs_path = Path(f'{pack}/{code_type}')
+            if code_dirs_path.exists():
+                for code_dir in code_dirs_path.glob('*'):
+                    cov_path = Path(f'{code_dir}/.coverage')
+                    if cov_path.exists():
+                        yield str(cov_path)
```
Comment (yaakovi) [inline demisto_sdk/commands/lint/helpers.py:484]:
```suggestion
    for cov_path in Path('Packs').glob('*/Integrations/*/.coverage'):
        yield str(cov_path)
    for cov_path in Path('Packs').glob('*/Scripts/*/.coverage'):
        yield str(cov_path)
```

File: demisto_sdk/commands/lint/helpers.py [hunk 3]
```diff
@@ -457,3 +459,48 @@ def split_warnings_errors(output: str):
...
+
+def generate_coverage_report(html=False, xml=False, report=True):
+    cov_dir = 'coverage_report'
+    cov_file = os.path.join(cov_dir, '.coverage')
+    cov = coverage.Coverage(data_file=cov_file)
+    cov.combine(coverage_files())
+    if os.path.exists(cov_file):
+        export_msg = 'exporting {0} coverage report to {1}'
+        if report:
+            print('############################\n unit-tests coverage report\n############################')
```
Comment (yaakovi) [inline demisto_sdk/commands/lint/helpers.py:495]: why a combination of `print` and `logger` is used here?

File: demisto_sdk/commands/lint/helpers.py [hunk 4]
```diff
@@ -457,3 +459,48 @@ def split_warnings_errors(output: str):
...
+                    if cov_path.exists():
+                        yield str(cov_path)
+
+
+def generate_coverage_report(html=False, xml=False, report=True):
+    cov_dir = 'coverage_report'
+    cov_file = os.path.join(cov_dir, '.coverage')
+    cov = coverage.Coverage(data_file=cov_file)
+    cov.combine(coverage_files())
+    if os.path.exists(cov_file):
```
Comment (yaakovi) [inline demisto_sdk/commands/lint/helpers.py:492]: swap the condition to reduce the function's "area under the indent"
```python
    if not os.path.exists(cov_file):
        logger.debug(f'skipping coverage report {cov_file} file not found.')
        return
```

**Review by yaakovi — COMMENTED** [review]
nice work.

---

we have 2 outputs for this command:
* text_xml - which the user can choose the path for it
* coverage report - which uses hard-coded path

I think it will be better if the user could control the output path of the coverage as well.
preferably, the user should set one output directory and enable which report he wants to generate by the command.

@roysagi - WDYT?

File: demisto_sdk/commands/lint/README.md [hunk 5]
```diff
@@ -50,17 +50,24 @@ Options:
...
 ---
 `demisto-sdk lint -g -p 2`
 1. lint and test check will execute on all Packages which are changed from `origin/master` and from in staging.
 2. 2 Threads will be used inorder to preform the lint.
 ---
+`demisto-sdk lint -i Integrations/PaloAltoNetworks_XDR,Scripts/HellowWorldScript --coverage`
+Details:
+1. lint and test check will execute on Packages `Integrations/PaloAltoNetworks_XDR,Scripts/HellowWorldScript`
+2. coverage report will be printed.
+3. coverage report files will be exported.
```
Comment (roysagi) [inline demisto_sdk/commands/lint/README.md:72]: good please add this to the content-docs as well.

File: demisto_sdk/commands/lint/README.md [hunk 6]
```diff
@@ -50,17 +50,28 @@ Options:
     Path to store json results
 *  **-lp, --log-path PATH**
     Path to store all levels of logs
+*  **--no-coverage**
+    do NOT report coverage
```
Comment (roysagi) [inline demisto_sdk/commands/lint/README.md:54]:
```suggestion
    Do NOT report coverage
```

File: demisto_sdk/commands/lint/helpers.py [hunk 7]
```diff
@@ -457,3 +459,60 @@ def split_warnings_errors(output: str):
...
+
+def coverage_files():
+    packs_pass = Path('Packs')
+    for cov_path in packs_pass.glob('*/Integrations/*/.coverage'):
+        yield str(cov_path)
+    for cov_path in packs_pass.glob('*/Scripts/*/.coverage'):
+        yield str(cov_path)
+
+
+def generate_coverage_report(html=False, xml=False, report=True, cov_dir='coverage_report'):
```
Comment (roysagi) [inline demisto_sdk/commands/lint/helpers.py:495]: please add a docstring here.

**Review by roysagi — APPROVED** [review]

**Comment by roysagi** [comment]
@esharf - Awsome work here.

**Pull request merged**